libzmix/src/utils/bls12_381/mod.rs

```diff
@@ -0,0 +1,237 @@
+use amcl::bls381::big::BIG;
+use amcl::bls381::dbig::DBIG;
+use amcl::bls381::ecp::ECP;
+use amcl::bls381::ecp2::ECP2;
+use amcl::bls381::rom;
+
+use sha2::Sha256;
+
+use rand::thread_rng;
+use rand_core::RngCore;
+
+use hkdf::Hkdf;
+
+lazy_static! {
+    static ref G1: ECP = ECP::generator();
+    static ref G2: ECP2 = ECP2::generator();
+    static ref GROUP_ORDER: BIG = BIG::new_ints(&rom::CURVE_ORDER);
+}
+
+macro_rules! serialize_impl {
+    ($name:ident, $internal:ident, $visitor:ident) => {
+        impl std::str::FromStr for $name {
+            type Err = std::num::ParseIntError;
+
+            fn from_str(s: &str) -> Result<Self, Self::Err> {
+                Ok($name($internal::from_hex(s.to_string())))
```

Comment:
Lovesh, your implementation does handle the case for non-hex input but it is not constant time. See https://github.com/lovesh/amcl_rust_wrapper/blob/master/src/field_elem.rs#L436. It bails early. There is also the if with a break. CS does not have any of these.

Reply:
It's not constant time only if the input is invalid or if the input has a smaller size than `MODBYTES*2` where it spends time padding the input. The `break` would always be executed during the last iteration of the loop to avoid an extra shift. And there are always `MODBYTES*2` iterations of the loop and `MODBYTES*2-1` 4-bit left shifts. So it is constant time for valid input and `MODBYTES*2` input of size `MODBYTES*2`

Reply:
That’s still not CS. However you only need constant time to protect secret values like when computing commitments or signing where a private key is used. Serialization I think of any kind doesn’t need to be constant time.

Reply:
It's constant time for the cases that matter. Serialization/Deserialization should be constant time. The attacker can learn about private key when you are loading a stored private key or serializing a private key for storage.